graphql-kotlin-federation/src/test/kotlin/com/expediagroup/graphql/federation/FederatedSchemaValidatorTest.kt

```diff
@@ -0,0 +1,226 @@
+            .withDirective(extendsDirectiveType)
+            .build()
+
+        assertFailsWith<InvalidFederatedSchema> {
+            FederatedSchemaValidator().validateGraphQLType(typeToValidate)
+        }
+    }
+
+    /**
+     * type Foo @extends {
+     *   bar: String
+     * }
+     */
+    @Test
+    fun `validate federated GraphQLObjectType with no key directive`() {
+        val field = GraphQLFieldDefinition.newFieldDefinition()
+            .name("bar")
+            .type(GraphQLString)
+
+        val typeToValidate = GraphQLObjectType.newObject()
+            .name("Foo")
+            .field(field)
+            .withDirective(extendsDirectiveType)
+            .build()
+
+        assertFailsWith<InvalidFederatedSchema> {
+            FederatedSchemaValidator().validateGraphQLType(typeToValidate)
+        }
+    }
+
+    /**
+     * type Foo @key {
+     *   bar: String
+     * }
+     */
+    @Test
+    fun `validate federated GraphQLObjectType with key directive`() {
+        val field = GraphQLFieldDefinition.newFieldDefinition()
+            .name("bar")
+            .type(GraphQLString)
+
+        val typeToValidate = GraphQLObjectType.newObject()
+            .name("Foo")
+            .field(field)
+            .withDirective(keyDirectiveType)
+            .build()
+
+        val result = kotlin.runCatching {
+            FederatedSchemaValidator().validateGraphQLType(typeToValidate)
+        }
+
+        val expectedError = """
+            Invalid federated schema:
+             - @key directive on Foo is missing field information
+        """.trimIndent()
+
+        assertEquals(expectedError, result.exceptionOrNull()?.message)
+    }
+
+    /**
+     * type Foo @key {
+     *   bar: String @provides
+     * }
+     */
+    @Test
+    fun `validate federated GraphQLObjectType with key directive and field with provides directive`() {
+        val directive = GraphQLDirective.newDirective().name("provides")
+        val field = GraphQLFieldDefinition.newFieldDefinition()
+            .name("bar")
+            .withDirective(directive)
+            .type(GraphQLString)
+
+        val typeToValidate = GraphQLObjectType.newObject()
+            .name("Foo")
+            .field(field)
+            .withDirective(keyDirectiveType)
+            .build()
+
+        val result = kotlin.runCatching {
+            FederatedSchemaValidator().validateGraphQLType(typeToValidate)
+        }
+
+        val expectedError = """
+            Invalid federated schema:
+             - @key directive on Foo is missing field information
+             - @provides directive is specified on a Foo.bar field but it does not return an object type
+        """.trimIndent()
+
+        assertEquals(expectedError, result.exceptionOrNull()?.message)
+    }
+
+    /**
+     * type Foo @key {
+     *   bar: String @requires
+     * }
+     */
+    @Test
+    fun `validate federated GraphQLObjectType with key directive and field with requires directive`() {
```

Comment:
should this be valid `@key` and invalid `@requires`?

Reply:
See above, done